README

		       Music Player Daemon (MPD)
			http://www.musicpd.org

A daemon for playing music of various formats.  Music is played through the
server's audio device.  The daemon stores info about all available music,
and this info can be easily searched and retrieved.  Player control, info
retrieval, and playlist management can all be managed remotely.

For basic installation information see the INSTALL file.

MPD is released under the GNU General Public License version 2, which is
distributed in the COPYING file.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.



NEWS
ver 0.17 (2011/??/??)
* protocol:
  - support client-to-client communication
  - "update" and "rescan" need only "CONTROL" permission
* input:
  - cdio_paranoia: new input plugin to play audio CDs
  - curl: enable CURLOPT_NETRC
  - ffmpeg: support libavformat 0.7
* decoder:
  - mpg123: implement seeking
  - ffmpeg: drop support for pre-0.5 ffmpeg
  - ffmpeg: support libavformat 0.7
* output:
  - osx: allow user to specify other audio devices
  - raop: new output plugin
  - shout: add possibility to set url
  - roar: new output plugin for RoarAudio
* state_file: add option "restore_paused"
* cue: show CUE track numbers


ver 0.16.3 (2011/??/??)
* fix assertion failure in audio format mask parser
* fix NULL pointer dereference in playlist parser
* fix playlist files in base music directory
* database: allow directories with just playlists
* decoder:
  - ffmpeg: support libavcodec 0.7


ver 0.16.2 (2011/03/18)
* configure.ac:
  - fix bashism in tremor test
* decoder:
  - tremor: fix configure test
  - gme: detect end of song
* encoder:
  - vorbis: reset the Ogg stream after flush
* output:
  - httpd: fix uninitialized variable
  - httpd: include sys/socket.h
  - oss: AFMT_S24_PACKED is little-endian
  - oss: disable 24 bit playback on FreeBSD


ver 0.16.1 (2011/01/09)
* audio_check: fix parameter in prototype
* add void casts to suppress "result unused" warnings (clang)
* input:
  - ffado: disable by default
* decoder:
  - mad: work around build failure on Solaris
  - resolve modplug vs. libsndfile cflags/headers conflict
* output:
  - solaris: add missing parameter to open_cloexec() cal
  - osx: fix up audio format first, then apply it to device
* player_thread: discard empty chunks while cross-fading
* player_thread: fix assertion failure due to early seek
* output_thread: fix double lock


ver 0.16 (2010/12/11)
* protocol:
  - send song modification time to client
  - added "update" idle event
  - removed the deprecated "volume" command
  - added the "findadd" command
  - range support for "delete"
  - "previous" really plays the previous song
  - "addid" with negative position is deprecated
  - "load" supports remote playlists (extm3u, pls, asx, xspf, lastfm://)
  - allow changing replay gain mode on-the-fly
  - omitting the range end is possible
  - "update" checks if the path is malformed
* archive:
  - iso: renamed plugin to "iso9660"
  - zip: renamed plugin to "zzip"
* input:
  - lastfm: obsolete plugin removed
  - ffmpeg: new input plugin using libavformat's "avio" library
* tags:
  - added tags "ArtistSort", "AlbumArtistSort"
  - id3: revised "performer" tag support
  - id3: support multiple values
  - ape: MusicBrainz tags
  - ape: support multiple values
* decoders:
  - don't try a plugin twice (MIME type & suffix)
  - don't fall back to "mad" unless no plugin matches
  - ffmpeg: support multiple tags
  - ffmpeg: convert metadata to generic format
  - ffmpeg: implement the libavutil log callback
  - sndfile: new decoder plugin based on libsndfile
  - flac: moved CUE sheet support to a playlist plugin
  - flac: support streams without STREAMINFO block
  - mikmod: sample rate is configurable
  - mpg123: new decoder plugin based on libmpg123
  - sidplay: support sub-tunes
  - sidplay: implemented songlength database
  - sidplay: support seeking
  - sidplay: play monaural SID tunes in mono
  - sidplay: play mus, str, prg, x00 files
  - wavpack: activate 32 bit support
  - wavpack: allow more than 2 channels
  - mp4ff: rename plugin "mp4" to "mp4ff"
* encoders:
  - twolame: new encoder plugin based on libtwolame
  - flac: new encoder plugin based on libFLAC
  - wave: new encoder plugin for PCM WAV format
* output:
  - recorder: new output plugin for recording radio streams
  - alsa: don't recover on CANCEL
  - alsa: fill period buffer with silence before draining
  - openal: new output plugin
  - pulse: announce "media.role=music"
  - pulse: renamed context to "Music Player Daemon"
  - pulse: connect to server on MPD startup, implement pause
  - jack: require libjack 0.100
  - jack: don't disconnect during pause
  - jack: connect to server on MPD startup
  - jack: added options "client_name", "server_name"
  - jack: clear ring buffers before activating
  - jack: renamed option "ports" to "destination_ports"
  - jack: support more than two audio channels
  - httpd: bind port when output is enabled
  - httpd: added name/genre/website configuration
  - httpd: implement "pause"
  - httpd: bind_to_address support (including IPv6)
  - oss: 24 bit support via OSS4
  - win32: new output plugin for Windows Wave
  - shout, httpd: more responsive to control commands
  - wildcards allowed in audio_format configuration
  - consistently lock audio output objects
* player:
  - drain audio outputs at the end of the playlist
* mixers:
  - removed support for legacy mixer configuration
  - reimplemented software volume as mixer+filter plugin
  - per-device software/hardware mixer setting
* commands:
  - added new "status" line with more precise "elapsed time"
* update:
  - automatically update the database with Linux inotify
  - support .mpdignore files in the music directory
  - sort songs by album name first, then disc/track number
  - rescan after metadata_to_use change
* normalize: upgraded to AudioCompress 2.0
  - automatically convert to 16 bit samples
* replay gain:
  - reimplemented as a filter plugin
  - fall back to track gain if album gain is unavailable
  - optionally use hardware mixer to apply replay gain
  - added mode "auto"
  - parse replay gain from APE tags
* log unused/unknown block parameters
* removed the deprecated "error_file" option
* save state when stopped
* renamed option "--stdout" to "--stderr"
* removed options --create-db and --no-create-db
* state_file: save only if something has changed
* database: eliminated maximum line length
* log: redirect stdout/stderr to /dev/null if syslog is used
* set the close-on-exec flag on all file descriptors

src/utils.h
/*
 * Copyright (C) 2003-2010 The Music Player Daemon Project
 * http://www.musicpd.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef MPD_UTILS_H
#define MPD_UTILS_H

#include <stdbool.h>

#ifndef assert_static
/* Compile time assertion developed by Ralf Holly */
/* http://pera-software.com/articles/compile-time-assertions.pdf */
#define assert_static(e) \
	do { \
		enum { assert_static__ = 1/(e) }; \
	} while (0)
#endif /* !assert_static */

char *parsePath(char *path);

#endif


src/mixer_list.h
/*
 * Copyright (C) 2003-2010 The Music Player Daemon Project
 * http://www.musicpd.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/** \file
 *
 * This header provides "extern" declarations for all mixer plugins.
 */

#ifndef MPD_MIXER_LIST_H
#define MPD_MIXER_LIST_H

extern const struct mixer_plugin software_mixer_plugin;
extern const struct mixer_plugin alsa_mixer_plugin;
extern const struct mixer_plugin oss_mixer_plugin;
extern const struct mixer_plugin pulse_mixer_plugin;
extern const struct mixer_plugin winmm_mixer_plugin;

#endif


src/decoder/gme_decoder_plugin.c
#include "config.h"
#include "../decoder_api.h"
#include "audio_check.h"
#include "uri.h"
#include <assert.h>
#include <errno.h>
#include <stdlib.h>
#include <glib.h>
#include <string.h>

#include <gme/gme.h>

#undef G_LOG_DOMAIN
#define G_LOG_DOMAIN "gme"

#define SUBTUNE_PREFIX "tune_"

enum {
	GME_SAMPLE_RATE = 44100,
	GME_CHANNELS = 2,
	GME_BUFFER_FRAMES = 2048,
	GME_BUFFER_SAMPLES = GME_BUFFER_FRAMES * GME_CHANNELS,
};

/**
 * returns the file path stripped of any /tune_xxx.* subtune
 * suffix
 */
static char *
get_container_name(const char *path_fs)
{
	const char *subtune_suffix = uri_get_suffix(path_fs);
	char *path_container=g_strdup(path_fs);
	char *pat = g_strconcat("*/" SUBTUNE_PREFIX "???.", subtune_suffix, NULL);
	GPatternSpec *path_with_subtune=g_pattern_spec_new(pat);
	g_free(pat);
	if(!g_pattern_match(path_with_subtune,
		strlen(path_container), path_container, NULL)){
		g_pattern_spec_free(path_with_subtune);
		return path_container;
    }

	char *ptr=g_strrstr(path_container, "/" SUBTUNE_PREFIX);
	if(ptr) *ptr='\0';

	g_pattern_spec_free(path_with_subtune);
	return path_container;
}

/**
 * returns tune number from file.nsf/tune_xxx.* style path or 0 if no subtune
 * is appended.
 */
static int
get_song_num(const char *path_fs)
{
    const char *subtune_suffix = uri_get_suffix(path_fs);
	char *pat = g_strconcat("*/" SUBTUNE_PREFIX "???.", subtune_suffix, NULL);
	GPatternSpec *path_with_subtune=g_pattern_spec_new(pat);
	g_free(pat);

	if(g_pattern_match(path_with_subtune,
		strlen(path_fs), path_fs, NULL)) {
		char *sub=g_strrstr(path_fs, "/" SUBTUNE_PREFIX);
		g_pattern_spec_free(path_with_subtune);
		if(!sub){
            return 0;
        }

		sub+=strlen("/" SUBTUNE_PREFIX);
		int song_num=strtol(sub, NULL, 10);

		return song_num-1;
	} else
		g_pattern_spec_free(path_with_subtune);
		return 0;

}

static char *
gme_container_scan(const char *path_fs, const unsigned int tnum)
{
	Music_Emu *emu;
	const char* gme_err;
    unsigned int num_songs;

	gme_err = gme_open_file(path_fs, &emu, GME_SAMPLE_RATE);
	if (gme_err != NULL) {
		g_warning("%s", gme_err);
		return NULL;
	}

	num_songs = gme_track_count(emu);
	/* if it only contains a single tune, don't treat as container */
    if(num_songs < 2)
		return NULL;

	const char *subtune_suffix = uri_get_suffix(path_fs);
	if(tnum <= num_songs){
		char *subtune = g_strdup_printf(
			SUBTUNE_PREFIX "%03u.%s", tnum, subtune_suffix);
		return subtune;
	} else
		return NULL;
}

static void
gme_file_decode(struct decoder *decoder, const char *path_fs)
{
	float song_len;
	Music_Emu *emu;
	gme_info_t *ti;
	struct audio_format audio_format;
	enum decoder_command cmd;
	short buf[GME_BUFFER_SAMPLES];
	const char* gme_err;
	char *path_container=get_container_name(path_fs);
    int song_num;

    song_num=get_song_num(path_fs);

	gme_err = gme_open_file(path_container, &emu, GME_SAMPLE_RATE);
	g_free(path_container);
	if (gme_err != NULL) {
		g_warning("%s", gme_err);
		return;
	}

	if((gme_err = gme_track_info(emu, &ti, song_num)) != NULL){
		g_warning("%s", gme_err);
		gme_delete(emu);
		return;
	}

	if(ti->length > 0)
		song_len = ti->length / 1000.0;
	else song_len = -1;

	/* initialize the MPD decoder */

	GError *error = NULL;
	if (!audio_format_init_checked(&audio_format, GME_SAMPLE_RATE,
				       SAMPLE_FORMAT_S16, GME_CHANNELS,
				       &error)) {
		g_warning("%s", error->message);
		g_error_free(error);
		gme_free_info(ti);
		gme_delete(emu);
		return;
	}

	decoder_initialized(decoder, &audio_format, true, song_len);

	if((gme_err = gme_start_track(emu, song_num)) != NULL)
		g_warning("%s", gme_err);

	/* play */
	do {
		gme_err = gme_play(emu, GME_BUFFER_SAMPLES, buf);
		if (gme_err != NULL) {
			g_warning("%s", gme_err);
			return;
		}
		cmd = decoder_data(decoder, NULL, buf, sizeof(buf), 0);

		if(cmd == DECODE_COMMAND_SEEK) {
			float where = decoder_seek_where(decoder);
			if((gme_err = gme_seek(emu, (int)where*1000)) != NULL)
				g_warning("%s", gme_err);
			decoder_command_finished(decoder);
		}

		if(gme_track_ended(emu))
			break;
	} while(cmd != DECODE_COMMAND_STOP);

	gme_free_info(ti);
	gme_delete(emu);
}

static struct tag *
gme_tag_dup(const char *path_fs)
{
	Music_Emu *emu;
	gme_info_t *ti;
	const char* gme_err;
	char *path_container=get_container_name(path_fs);
	int song_num;
	song_num=get_song_num(path_fs);

	gme_err = gme_open_file(path_container, &emu, GME_SAMPLE_RATE);
	g_free(path_container);
	if (gme_err != NULL) {
		g_warning("%s", gme_err);
		return NULL;
	}
	if((gme_err = gme_track_info(emu, &ti, song_num)) != NULL){
		g_warning("%s", gme_err);
		gme_delete(emu);
		return NULL;
	}

	struct tag *tag = tag_new();
	if(ti != NULL){
		if(ti->length > 0)
			tag->time = ti->length / 1000;
		if(ti->song != NULL){
			if(gme_track_count(emu) > 1){
				/* start numbering subtunes from 1 */
				char *tag_title=g_strdup_printf("%s (%d/%d)",
					ti->song, song_num+1, gme_track_count(emu));
				tag_add_item(tag, TAG_TITLE, tag_title);
				g_free(tag_title);
			}else
				tag_add_item(tag, TAG_TITLE, ti->song);
		}
		if(ti->author != NULL)
			tag_add_item(tag, TAG_ARTIST, ti->author);
		if(ti->game != NULL)
			tag_add_item(tag, TAG_ALBUM, ti->game);
		if(ti->comment != NULL)
			tag_add_item(tag, TAG_COMMENT, ti->comment);
		if(ti->copyright != NULL)
			tag_add_item(tag, TAG_DATE, ti->copyright);
	}

	gme_free_info(ti);
	gme_delete(emu);
	return tag;
}

static const char *const gme_suffixes[] = {
	"ay", "gbs", "gym", "hes", "kss", "nsf",
	"nsfe", "sap", "spc", "vgm", "vgz",
	NULL
};

extern const struct decoder_plugin gme_decoder_plugin;
const struct decoder_plugin gme_decoder_plugin = {
	.name = "gme",
	.file_decode = gme_file_decode,
	.tag_dup = gme_tag_dup,
	.suffixes = gme_suffixes,
    .container_scan = gme_container_scan,
};


src/decoder/adplug_decoder_plugin.cxx
#include "config.h"

extern "C" {
#include "../decoder_api.h"
#include "audio_check.h"
}

#include <glib.h>
//#include <assert.h>
#include <adplug/adplug.h>
#include <adplug/emuopl.h>
#include <adplug/kemuopl.h>

#undef G_LOG_DOMAIN
#define G_LOG_DOMAIN "adplug"

#ifndef min
#  define min(a,b) (((a) < (b)) ? (a) : (b))
#endif


enum {
	ADPLUG_SAMPLE_RATE = 44100,
	ADPLUG_CHANNELS = 1,
    ADPLUG_BUFFER_FRAMES = 256,
    ADPLUG_BUFFER_SAMPLES = ADPLUG_BUFFER_FRAMES * ADPLUG_CHANNELS,
};

inline unsigned char getsampsize() { return (ADPLUG_CHANNELS * (16 / 8)); }

static void
adplug_file_decode(struct decoder *decoder, const char *path_fs)
{
    /* don't know the difference between these two emulators,
     * should it be made configurable?
     */
    Copl *opl = new CEmuopl(44100, true, false);
    //Copl *opl = new CKemuopl(44100, true, true);

    /* TODO the mpd log gets cluttered up with a bunch of messages from this
     * constructor! */
    CPlayer *p = CAdPlug::factory(path_fs, opl);
	short buf[ADPLUG_BUFFER_SAMPLES];
    if(!p){
        g_warning("Unknown filetype\n");
        return;
    }
    /* no support for subsongs yet */
    p->rewind(0);

	/* initialize the MPD decoder */
	struct audio_format audio_format;
	enum decoder_command cmd;

	GError *error = NULL;
	if (!audio_format_init_checked(&audio_format, ADPLUG_SAMPLE_RATE,
				       SAMPLE_FORMAT_S16, 2,
				       &error)) {
		g_warning("%s", error->message);
		g_error_free(error);
		return;
	}

	decoder_initialized(decoder, &audio_format, true, NULL);
    /* play */
    do {
        static long	minicnt = 0;
        long		i, towrite = ADPLUG_BUFFER_SAMPLES / getsampsize();
        short		*pos = buf;

        // Prepare audiobuf with emulator output
        while(towrite > 0) {
            while(minicnt < 0) {
                  minicnt += ADPLUG_SAMPLE_RATE;
                  //self->playing = self->p->update();
                  p->update();
            }
            i = min(towrite, (long)(minicnt / p->getrefresh() + 4) & ~3); //the fuck??!
            g_log(G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG, "li: %i", i);
            opl->update((short *)pos, i);
            pos += i * getsampsize();
            towrite -= i;
            minicnt -= (long)(p->getrefresh() * i);
        }
        cmd = decoder_data(decoder, NULL, buf, sizeof(buf), 0);
    } while(cmd != DECODE_COMMAND_STOP);

}

static struct tag *
adplug_tag_dup(const char *path_fs)
{
    /* don't know the difference between these two emulators,
     * should it be made configurable?
     */
    Copl *opl = new CEmuopl(44100, true, true);
    //Copl *opl = new CKemuopl(44100, true, true);
    CPlayer *p = CAdPlug::factory(path_fs, opl);
    if(!p){
        g_warning("Unknown filetype\n");
        return NULL;
    }
	struct tag *tag = tag_new();
    tag_add_item(tag, TAG_ARTIST, p->getauthor().c_str());
    tag_add_item(tag, TAG_TITLE, p->gettitle().c_str());
    return tag;
}

static const char *const adplug_suffixes[] = {
    "a2m", "adl", "adlib", "amd", "bam", "cff", "cmf", "d00", "dtm", "dfm",
    "dmo", "dro", "hsc", "hsp", "imf", "laa", "lds", "m", "mad", "mid", "mkj",
    "msc", "mtk", "rad", "raw", "rix", "rol", "s3m", "sa2", "sat", "sci", "sng",
    "wlf", "xad", "xsm",
    NULL
};

extern const struct decoder_plugin adplug_decoder_plugin;
const struct decoder_plugin adplug_decoder_plugin = {
    "adplug",
    NULL, /* init() */
    NULL, /* finish() */
    NULL, /* stream_decode() */
    adplug_file_decode,
    adplug_tag_dup, /* tag_dup() */
    NULL, /* stream_tag() */
    NULL, /* container_scan() */
    adplug_suffixes,
    NULL, /* mime_types */
};
